CompositeJK Part 2: Pilot Implementation (#2833)
psi4/psi4: doc/sphinxman/source/scf.rst

The key difficulty in the SCF procedure is treatment of the four-index ERI
contributions to the Fock Matrix. A number of algorithms are available in
|PSIfour| for these terms. The algorithm is selected by the |globals__scf_type|

> keyword. Some such algorithms consist of a single algorithm applied to

Author: Changes in doc/* files are documentation updates regarding CompositeJK - worth looking at.

> keyword. Some such algorithms consist of a single algorithm applied to
> the consutruction of both the Coulomb and Exchange constructions
> to the Fock Matrix, such as follows:

Reviewer: ```suggestion
keyword. Most consist of a single algorithm applied to
the construction of both the Coulomb and Exchange parts
of the Fock Matrix:
```

    for gradient computations.  The algorithm to obtain the Cholesky
    vectors is not designed for computations with thousands of basis
    functions.

|PSIfour| also features the capability to use arbitrary combinations 
of specialized algorithms that construct either the Coulomb

> or the Exchange matrix separately. These algorithms can be mixed and
> matched arbitrarily to optimize performance based on the calculation
> in question. In general, such combinations of algorithms display lower
> scaling then their combined counterparts, and thus perform better 
> with larger systems. Such combinations of algorithms can be utilized 
> by setting the |globals__scf_type| keyword to ``J_alg+K_alg``, 
> where *J_alg* is the name of the separate Coulomb construction algoritm to use, 
> and *K_alg* is the name of the separate Exchange construction algorithm to use.

Reviewer: ```suggestion
or the Exchange matrix separately. In general, such combinations of algorithms display lower
scaling than their combined counterparts and thus perform better 
for larger systems. Such combinations can be accessed 
by setting |globals__scf_type| to ``J_alg+K_alg``, 
where *J_alg* and *K_alg* are the names of the separate Coulomb
and Exchange construction algorithms, respectively.
```

necessary for combined Fock build algorithms. The overall consequence, then, is that composite combinations of separate J and K construction 
algorithms tend to perform better than combined Fock build algorithms for sufficiently large system sizes, while performing worse for smaller systems. 

Multiple separate J and K construction algorithms to construct the Fock matrix in a composite fashion are available for use in Psi4, and they can be 
mixed and matched with each other freely. The algorithms availble for use in a composite fashion are as follows:

Separate J Construction: DIRECTDFJ 

Separate K Construction: COSX, LINK

Two composite algorithms can be combined with each other by setting the |globals__scf_type| keyword to ``J_alg+K_alg``, where *J_alg* is an available
algorithm for the separate construction of J, and *K_alg* is an available algorithm for the separate construction of K. For example, if one wanted
to run an integral-direct density-fitted algorithm for Coulomb construction (i.e., DIRECTDFJ) in conjunction with the Linear Exchange algorithm

> for Exchange construction (i.e., LINK), one would use the |globals__scf_type| keyword ``DIRECTDFJ+LINK``.  

Reviewer: Isn't this (sec:compositejk) pretty repetitive from the new part further up?

Author: Hmm... I suppose one could come to that conclusion. There are a couple of details mentioned in `sec:compositejk` that aren't mentioned in the CompositeJK summary (i.e. the "new part further up"), such as the exploitation of different properties in J and K for composite algorithms, the existence of redundant ERI computation in composite algorithms, and composite algorithms performing worse for small systems (the latter is implied in the summary, but not explicitly stated). How important these details are, and whether they could simply be bundled into the summary, is the question.

Reviewer: I vote to bundle the redundant ERI, worse for large, diff J, K properties, etc. details into the summary. That is, have one big section or a little intro section and a big section, rather than the two big sections there are now.
